### ISSUE TITLE
Improve checkout feature

### ISSUE THREAD

**Issue by rafsanulhasan** [smartstore/SmartStoreNET#1786]
Hi maintainer,
I'm trying to integrate SSLCommerz (sslcommerz.com) payment gateway. To do that, I'm trying to develop a plugin. I'm facing some problem while integrating the payment method.

SSL Commerz is a  payment gateway which supports multiple channels like VISA, AMEX, MASTERCARD and other mobile banking and so on. It is required to send a "**transaction id**" while initiating a payment initially in order for the gateway to track the transaction by its id because when the gateway makes a POST request to our IPNHandler, it sends back the transaction id using which we should update the _order_ in our database.

**Comment by mgesing** [comment]
> Is there any other way to get the order id prior to make the payment initiation request?

No. Use [ProcessPaymentRequest.OrderGuid](https://github.com/smartstore/SmartStoreNET/blob/4.x/src/Plugins/SmartStore.AmazonPay/Services/AmazonPayService.cs#L960) instead. It is stored in the database at Order.OrderGuid after the order has been placed.

> Proposed: Cart->Addresses->Shipping method->Confirm->Order Complete->Payment->Payment complete->Invoice shown/sent

This is possible but too extensive to explain here. Basically, you need to change the checkout flow through an IActionFilter implementation. Please take a closer look at the [AmazonPay](https://github.com/smartstore/SmartStoreNET/blob/4.x/src/Plugins/SmartStore.AmazonPay/Filters/AmazonPayCheckoutFilter.cs#L9) plugin because it does something similar.

**Comment by rafsanulhasan** [comment]
@mgesing How could I make use of ProcessPayment request. In the controller which is inherited from PaymentControllerbase?

**Comment by mgesing** [comment]
You create a payment provider that inherits from PaymentMethodBase and overwrites the methods required for payment processing (like ProcessPayment). The [PayPal](https://github.com/smartstore/SmartStoreNET/tree/4.x/src/Plugins/SmartStore.PayPal/Providers) plugin has several providers that illustrate this.

**Comment by rafsanulhasan** [comment]
@mgesing I gave it a try.
I implemented a provider (inherits from paymentproviderbase), implemented controller (paymentcontrollerbase).

**Comment by mgesing** [comment]
You have to install the plugin and activate the payment method under Configuration > Payment methods.

**Comment by rafsanulhasan** [comment]
Ya I did it and now it works correctly but still it does not behave as expected. I wonder how to handle the case: 

when the customer has paid the money but failed to create the order successfully because for some reason customer browser restarted.

I think the second point I was talking above is the only solution to the problem. Changing the checkout process flow could easily solve the problem. if the order is created prior to make the payment, the order is created with its status pending and payment status is also pending. then when the user pays the invoiced amount, the record is updated. in case of user browser of machine shuts down, the order is still pending until successful payment and wouldn't be processed until making making payment.

Now the problem is, can I really change the checkout process flow as done in AmazonPay plugin? How could it be changed with a plugins. I'm asking this because I wonder 

1. how could a plugin can change the process flow while the other plugins work well with the existing flow.
2. before selecting the plugin in the PaymentMethod step, how could the system know that that a specific plugin, that uses a different flow, is being used?

**Comment by mgesing** [comment]
Collecting money from shop visitors in a way that may result in missing orders is a basically wrong payment implementation or checkout integration.

Amazon Pay cannot be started from the payment method list. This would be too late regarding the checkout flow. It can only be started from the shopping cart page or the flyout shopping cart dialog. This way the software knows, from here on only Amazon Pay, other payment method only by going back to the shopping cart page and reentering the checkout..

**Comment by rafsanulhasan** [comment]
@mgesing is there any way to publish plugins to your marketplace via CI/CD

**Comment by mgesing** [comment]
No, I guess not.
